File: data/dtos.py
from typing import Any

from pydantic import BaseModel, Field


class MessageDTO(BaseModel):
    role: str = Field(
        ..., description="The role of the message sender (user/assistant/system)"
    )
    content: str = Field(..., description="The content of the message")


class ResponseDTO(BaseModel):
    tokens_used: int = Field(
        ...,
        description="Number of tokens used in the response",
    )
    model: str = Field(
        ...,
        description="The model used for generating the response",
    )
    finish_reason: str | None = Field(
        None,
        description="Reason why the response generation finished",
    )
    raw_response: dict[str, Any] | None = Field(
        None,
        description="Raw response from the LLM API",
    )

    def get_response_type(self) -> str:
        return "base"


class TextResponseDTO(ResponseDTO):
    content: str = Field(..., description="The generated text content")

    def get_response_type(self) -> str:
        return "text"


class FunctionCallResponseDTO(ResponseDTO):
    function_name: str = Field(
        ...,
        description="The name of the function to be called",
    )
    function_arguments: dict[str, Any] = Field(
        ...,
        description="Arguments for the function call",
    )

    def get_response_type(self) -> str:
        return "function_call"

File: src/data/dtos.ts
/**
 * Data Transfer Objects for the data layer.
 */

export interface MessageDTO {
  readonly role: string;
  readonly content: string;
}

export interface BaseResponseDTO {
  readonly tokensUsed: number;
  readonly model: string;
  readonly finishReason?: string;
  readonly rawResponse?: Record<string, unknown>;
}

export interface TextResponseDTO extends BaseResponseDTO {
  readonly type: 'text';
  readonly content: string;
}

export interface FunctionCallResponseDTO extends BaseResponseDTO {
  readonly type: 'function_call';
  readonly functionName: string;
  readonly functionArguments: Record<string, unknown>;
}

export type ResponseDTO = TextResponseDTO | FunctionCallResponseDTO;

export function isTextResponseDTO(dto: ResponseDTO): dto is TextResponseDTO {
  return dto.type === 'text';
}

export function isFunctionCallResponseDTO(dto: ResponseDTO): dto is FunctionCallResponseDTO {
  return dto.type === 'function_call';
}
